### PR TITLE
Training loop

### PR DESCRIPTION
**Pull request by AdvaitGadhikar**
Hi Everyone, 

This PR adds a training loop and test file for the CIFAR-10 dataset like the files in optimization/cifar100.

Thank You!

### CODE REVIEW

**Comment by nightldj** [comment]
Thanks. Could you change the cla status? If you have signed the CLA before, maybe add a comment @cla and saying 'I signed it!'. 

And could you pull the current version of cifar10 in repo to resolve the conflicts? Thanks.

**Comment by AdvaitGadhikar** [comment]
@cla I signed it

**Comment by AdvaitGadhikar** [comment]
@googlebot I consent

**Comment by nightldj** [comment]
@JYWa It shows that we need you to sign it. Maybe also add a comment of "I consent" here? Thanks!

**Comment by JYWa** [comment]
@googlebot I consent

File: fedopt_guide/cifar10/centralized_cifar10.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+from utils.datasets import cifar10_dataset
+from utils.models import resnet_models
+
+
+CIFAR_SHAPE = (32, 32, 3)
+NUM_CHANNELS = 3
+NUM_CLASSES = 100
+
+
+def run_centralized(optimizer: tf.keras.optimizers.Optimizer,
```
Comment (zcharles8) [inline fedopt_guide/cifar10/centralized_cifar10.py:30]: Can't we simply use tensorflow datasets to load CIFAR-10 here? It seems wasteful to load it, partition, then re-amalgamate.

File: fedopt_guide/cifar10/federated_cifar10.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+"""Federated CIFAR-10 classification library using TFF."""
+
+import functools
+from typing import Callable, Optional
+
+from absl import logging
+import tensorflow as tf
+import tensorflow_federated as tff
+
+from utils import training_loop
```
Comment (zcharles8) [inline fedopt_guide/cifar10/federated_cifar10.py:23]: Let's use the training loop in the fedopt_guide folder. It's designed specifically for these tasks.

File: fedopt_guide/cifar10/federated_cifar10.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  We assume that the iterative process has the following functional type
+  signatures:
+    *   `initialize`: `( -> S@SERVER)` where `S` represents the server state.
+    *   `next`: `<S@SERVER, {B*}@CLIENTS> -> <S@SERVER, T@SERVER>` where `S`
+        represents the server state, `{B*}` represents the client datasets,
+        and `T` represents a python `Mapping` object.
+  The iterative process must also have a callable attribute `get_model_weights`
+  that takes as input the state of the iterative process, and returns a
+  `tff.learning.ModelWeights` object.
+  Args:
```
Comment (zcharles8) [inline fedopt_guide/cifar10/federated_cifar10.py:57]: Line break

File: fedopt_guide/cifar10/federated_cifar10.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    client_batch_size: int,
+    clients_per_round: int,
+    client_datasets_random_seed: Optional[int] = None,
+    crop_size: Optional[int] = 24,
+    total_rounds: Optional[int] = 1500,
+    experiment_name: Optional[str] = 'federated_cifar10',
+    root_output_dir: Optional[str] = '/tmp/fed_opt',
+    **kwargs):
+  """Runs an iterative process on the CIFAR-10 classification task.
+  This method will load and pre-process dataset and construct a model used for
```
Comment (zcharles8) [inline fedopt_guide/cifar10/federated_cifar10.py:44]: Line break

File: fedopt_guide/cifar10/federated_cifar10.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      clients_per_round=clients_per_round,
+      random_seed=client_datasets_random_seed)
+
+  test_fn = training_utils.build_centralized_evaluate_fn(
+      eval_dataset=cifar_test,
+      model_builder=model_builder,
+      loss_builder=loss_builder,
+      metrics_builder=metrics_builder)
+
+  validation_fn = lambda model_weights, round_num: test_fn(model_weights)
```
Comment (zcharles8) [inline fedopt_guide/cifar10/federated_cifar10.py:122]: A necessary follow-up would be to allow a federated evaluation instead (depending on what the user chooses) which would allow us to aggregate client-level evaluation statistics (eg. minimum accuracy across clients).

File: fedopt_guide/cifar10/federated_trainer.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+  flags.DEFINE_integer(
+      'rounds_per_eval', 1,
+      'How often to evaluate the global model on the validation dataset.')
+  flags.DEFINE_integer('rounds_per_checkpoint', 50,
+                       'How often to checkpoint the global model.')
+  flags.DEFINE_integer(
+      'rounds_per_profile', 0,
+      '(Experimental) How often to run the experimental TF profiler, if >0.')
+
+with utils_impl.record_hparam_flags() as task_flags:
```
Comment (zcharles8) [inline fedopt_guide/cifar10/federated_trainer.py:73]: This seems like an unnecessary holdover from the federated trainer in optimization/.

File: fedopt_guide/cifar10/federated_trainer.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        is the total number of examples processed on device.
+    Returns:
+      A `tff.templates.IterativeProcess`.
+    """
+
+    return tff.learning.build_federated_averaging_process(
+        model_fn=model_fn,
+        client_optimizer_fn=client_optimizer_fn,
+        server_optimizer_fn=server_optimizer_fn,
+        client_weight_fn=client_weight_fn,
```
Comment (zcharles8) [inline fedopt_guide/cifar10/federated_trainer.py:164]: I believe this has changed to client_weighting in tff-nightly

File: fedopt_guide/cifar10/federated_trainer.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        server_optimizer_fn=server_optimizer_fn,
+        client_weight_fn=client_weight_fn, 
+        use_experimental_simulation_loop=True)
+
+  shared_args = utils_impl.lookup_flag_values(shared_flags)
+  shared_args['iterative_process_builder'] = iterative_process_builder
+  task_args = _get_task_args()
+  hparam_dict = _get_hparam_flags()
+
+  if FLAGS.task == 'cifar10':
```
Comment (zcharles8) [inline fedopt_guide/cifar10/federated_trainer.py:172]: Seems unnecessary, since this is purely for cifar10.

File: utils/datasets/cifar10_dataset_test.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
+# Copyright 2021, Google LLC.
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset_test.py:1]: This file is already in the github. Is this diff up to date?

File: utils/datasets/cifar10_dataset.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
+# Copyright 2021, Google LLC.
```
Comment (zcharles8) [inline utils/datasets/cifar10_dataset.py:1]: This file is also in the github currently. Maybe the diff isn't up to date (not an expert on Github reviewing)?

**Review by zcharles8 — CHANGES_REQUESTED** [review]

File: fedopt_guide/cifar10/centralized_cifar10.py (hunk 1 above)
Comment (AdvaitGadhikar) [reply fedopt_guide/cifar10/centralized_cifar10.py:30]: Yes that makes sense, it would definitely become a lot simpler!

File: utils/datasets/cifar10_dataset.py (hunk 10 above)
Comment (AdvaitGadhikar) [reply utils/datasets/cifar10_dataset.py:1]: There were a couple of merge conflicts in these files because of the spacing, so I had to modify them to resolve the conflicts. I'm not very sure how to go about it either, any suggestions?

File: fedopt_guide/cifar10/federated_cifar10.py (hunk 5 above)
Comment (AdvaitGadhikar) [reply fedopt_guide/cifar10/federated_cifar10.py:122]: In this case would the test data also be distributed in the same non iid manner across clients?

File: fedopt_guide/cifar10/federated_trainer.py (hunk 7 above)
Comment (AdvaitGadhikar) [reply fedopt_guide/cifar10/federated_trainer.py:164]: Just a general question, should I always refer to the nightly version?

**Review by zcharles8 — APPROVED** [review]

**Review by nightldj — APPROVED** [review]

**Comment by nightldj** [comment]
Could you pull/sync as there seems to be some conflicts that prevent me from merging? Thanks!

**Comment by AdvaitGadhikar** [comment]
Hi,

I am not able to find the conflicts in the PR, could you point me to the
conflicts so that I can correct them?

Sincerely
Advait

**Review by nightldj — APPROVED** [review]

File: utils/datasets/BUILD [hunk 11]
```diff
@@ -73,7 +81,8 @@ py_test(
     srcs = ["stackoverflow_word_prediction_test.py"],
     python_version = "PY3",
     srcs_version = "PY3",
-    deps = [":stackoverflow_word_prediction"],
+    deps = [":stackoverflow_word_predi
+ction"],
```
Comment (nightldj) [inline utils/datasets/BUILD:85]: Why do you need to change this file. I think it is complaining about this new line. Could you discard your change in utils/dataset/? Thanks!

**Review by nightldj — CHANGES_REQUESTED** [review]

File: utils/datasets/BUILD (hunk 11 above)
Comment (AdvaitGadhikar) [reply utils/datasets/BUILD]: Sorry about that! Made the correction.

**Review by nightldj — APPROVED** [review]
